src/main/java/org/testng/internal/PackageUtils.java

```diff
@@ -83,68 +84,92 @@ public static void addClassLoader(final ClassLoader loader) {
       }
     }
 
+    ParamContainer container = new ParamContainer(included,excluded,vResult);
+
     for (URL url : dirs) {
-      String protocol = url.getProtocol();
-      if (! matchTestClasspath(url, packageDirName, recursive)) {
+      packageName = processURL(packageName, container, packageOnly, recursive, packageDirName, url);
+    }
+    return vResult.toArray(new String[vResult.size()]);
+  }
+
+  private static String processURL(String pkgName, ParamContainer container,
+      String packageOnly, boolean recursive , String packageDirName, URL url) throws IOException {
+    String packageName = pkgName;
+    List<String> included = container.getIncluded();
+    List<String> excluded = container.getExcluded();
+    List<String> vResult = container.getResult();
+
+    String protocol = url.getProtocol();
+    if (! matchTestClasspath(url, packageDirName, recursive)) {
+      return packageName;
+    }
+
+    if ("file".equals(protocol)) {
```

Comment:
I chose to leave this as is, because with switch I also had to define an empty `default` else again sonarcube would crib.